File: backend/AMD.AutoService.GaragePro.Domain/Entities/Attachment.cs
namespace AMD.AutoService.GaragePro.Domain.Entities;

/// <summary>
/// ไฟล์แนบ — ลายเซ็น รูปรับรถ รูปตรวจเช็ค รูปก่อน/หลัง
///
/// [BIZ] DB เก็บเฉพาะ metadata + path · ตัวไฟล์อยู่บน storage
/// ประเมินไว้ ~25–30 ไฟล์ต่องาน (docs/02-domain-model.md) จึงห้ามเก็บ blob ลง DB
/// </summary>
public class Attachment
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public Guid JobId { get; set; }
    public Job? Job { get; set; }

    /// <summary>signature | intake | inspection | repair-before | repair-after | qc | document</summary>
    public string Kind { get; set; } = string.Empty;

    /// <summary>เอกสาร/รายการที่ไฟล์นี้ผูกอยู่ เช่น QuotationId</summary>
    public Guid? EntityId { get; set; }

    public string FileName { get; set; } = string.Empty;
    public string ContentType { get; set; } = string.Empty;
    public long SizeBytes { get; set; }

    /// <summary>path ใต้ storage root — ไม่เก็บ absolute path เพื่อให้ย้าย storage ได้</summary>
    public string RelativePath { get; set; } = string.Empty;

    /// <summary>ตรวจไฟล์ซ้ำและยืนยันความถูกต้องหลังย้าย storage</summary>
    public string? Sha256 { get; set; }

    public long UploadedByUserId { get; set; }
    public string UploadedByName { get; set; } = string.Empty;
    public DateTime UploadedAt { get; set; } = DateTime.UtcNow;
}

File: mobile/lib/models/attachment.dart
/// ไฟล์แนบของจ๊อบ — ตรงกับ AttachmentDto ใน Application/Attachments/AttachmentService.cs
library;

class Attachment {
  Attachment({
    required this.id,
    required this.kind,
    this.entityId,
    required this.fileName,
    required this.contentType,
    required this.sizeBytes,
    required this.relativePath,
    required this.uploadedByName,
    required this.uploadedAt,
  });

  final String id;
  final String kind;
  final String? entityId;
  final String fileName;
  final String contentType;
  final int sizeBytes;

  /// path ที่ใช้กับ GET /attachments/file?path= — **ต้องแนบ Bearer เสมอ**
  final String relativePath;

  final String uploadedByName;
  final DateTime uploadedAt;

  bool get isImage => contentType.startsWith('image/');

  static Attachment fromJson(Map<String, dynamic> j) => Attachment(
        id: j['id'] as String,
        kind: j['kind'] as String? ?? '',
        entityId: j['entityId'] as String?,
        fileName: j['fileName'] as String? ?? '',
        contentType: j['contentType'] as String? ?? '',
        sizeBytes: (j['sizeBytes'] as num?)?.toInt() ?? 0,
        relativePath: j['relativePath'] as String,
        uploadedByName: j['uploadedByName'] as String? ?? '',
        uploadedAt: DateTime.parse(j['uploadedAt'] as String).toLocal(),
      );
}

/// kind ที่ server ยอมรับ — AttachmentService.AllowedKinds
/// ส่ง kind อื่นจะถูกปฏิเสธด้วย ATTACHMENT_KIND_INVALID
abstract final class AttachmentKind {
  static const signature = 'signature';
  static const intake = 'intake';
  static const inspection = 'inspection';
  static const repairBefore = 'repair-before';
  static const repairAfter = 'repair-after';
  static const qc = 'qc';
  static const document = 'document';
  static const handoverSignature = 'handover-signature';
  static const chat = 'chat';
}

/// ข้อจำกัดฝั่ง server (AttachmentOptions) — ตรวจที่ client ก่อนเพื่อไม่ให้เสียเวลาอัปโหลดฟรี
abstract final class AttachmentLimits {
  static const maxSizeBytes = 15 * 1024 * 1024;
  static const allowedContentTypes = <String>{
    'image/png',
    'image/jpeg',
    'image/webp',
    'application/pdf',
  };
}
